### PR TITLE
extract code path shared between FromIterator and Extend

### PR DESCRIPTION
**Pull request by the8472**
This should eliminate the call to `reserve` on the `FromIterator` code path since that's redundant after the `with_capacity` call. So it should result in reduced IR.

I hope this will result in better compile times by itself, but if not it may still improve the numbers combined with the revert in #83797 which would increase the code weight of `reserve()`

@rustbot label T-libs-impl

### CODE REVIEW

File: library/alloc/src/vec/mod.rs [hunk 1]
```diff
@@ -2512,6 +2512,29 @@ impl<T, A: Allocator> Vec<T, A> {
...
+    ///
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `self` has sufficient spare capacity
+    /// to hold the items returned by the iterator.
+    ///
+    /// The bound `I: TrustedLen` ensures that the caller can safely know
+    /// how much needs to be allocated.
+    unsafe fn extend_prealloc_trustedlen<I: TrustedLen<Item = T>>(&mut self, iterator: I) {
```
Comment (bjorn3) [inline library/alloc/src/vec/mod.rs:2525]:
```suggestion
    unsafe fn extend_prealloc_trusted_len<I: TrustedLen<Item = T>>(&mut self, iterator: I) {
```

**Comment by the8472** [comment]
ping @kennytm

**Comment by JohnTitor** [comment]
Has something been changed after reverting #83770? I'm seeing https://github.com/rust-lang/rust/pull/83770#issuecomment-849531641 and would like to know if we should re-run a perf-check.

**Pull request closed without merging**